### PR TITLE
Fixing issues related to slashes in version slugs

### PR DESCRIPTION
**Pull request by gregmuellegger**
As documented in #1176 and  #1352, we have problems with slashes in slugs. At the moment, the slug do not enforce any character scheme, so that's why slashes can add up in there.

I do not recommend adjusting the url patterns to allow slashes as it gets difficult to avoid shadowing of url definitions that come later in the urls.py.

I suggest having a `VersionSlug` field that takes care of validating and normalizing version numbers and checking that they are unique in a project (as currently enforced by the `Version._meta.unique_together`).

Things to do:

* [x] Make sure that the same regex is used in the urls.py for version slug matching (currently a few different patterns are in use where some match something different).
* [x] Create failing test to get evidence that slashes in slugs are a problem.
* [x] Implement `VersionSlug` field with validation and noramlizing
* [x] Add migration to transform current invalid slugs in the DB (containing slashes) to normalized slugs
* [x] Support displaying the correct identifier/tagname/branch in `get_vcs_version_slug`.

### CODE REVIEW

**Comment by agjohnson** [comment]
I'd like to avoid altering the URL as well for this case. I think it's probably a good plan in general. What are your thoughts on slugging/detecting version `foo/3.1` vs `foo-3.1`? This would be an edge case maybe, because a collision would only happen for projects with both branches.

Performing a `urlencode` on the version would be ugly, but workable -- maybe a different encoding mechanism here?

File: readthedocs/builds/version_slug.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# Regex breakdown:
+#   \w -- start with alphanumeric value
+#   [-._\w] -- allow dash, dot, underscore, digit, lowercase ascii
+#   +? -- allow multiple of those, but be not greedy about the matching
+#   (?: ... ) -- wrap everything so that the pattern cannot escape when used in
+#                regexes.
+VERSION_SLUG_REGEX = '(?:\w[-._\w]+?)'
+
+
+version_slug_regex = re.compile(VERSION_SLUG_REGEX)
```
Comment (ericholscher) [inline readthedocs/builds/version_slug.py]: Is there a reason to have both the upper case and lower case ones defined? Will be confusing on import, and should likely just be one line with the compiled one, unless there's a reason not to?

**Comment by ericholscher** [comment]
I think we likely need to clean the slug names. If we just urlencode the slugs in this case, what will we do for documentation pages, etc? We really need to not have slashes in slug names at all, because it will break other places downstream.

**Comment by gregmuellegger** [comment]
I would convert slashes to be double underscores, like `v/1.0` -> `v__1.0`, that should be very uncommon in tag names. If that clashes then I would start appending `_a`, if that fails `_b` etc.
I think adding the commonly used "slug-extend" `-1` is no good because dashes and digits are too common in tagnames.

File: readthedocs/builds/version_slug.py (hunk 1 above)
Comment (gregmuellegger) [inline readthedocs/builds/version_slug.py]: The reason for the uncompiled one is to be able to integrate them into the `urlconf`. However I just found out that you can use the `.pattern` attribute on compiled regexes to get down to the pattern again.

So I will remove the uncompiled one.

**Comment by gregmuellegger** [comment]
@ericholscher while trying to write the migration for the broken version slugs I noticed the `VersionAlias` model. It seems that there is only code that allows to edit the alias in the admin, but the aliases are never used in the frontend somehow. Is that right? Is the `VersionAlias` model and everything related code that we can remove?

**Comment by gregmuellegger** [comment]
I added `VersionSlugField` and the migration necessary to normalize currently invalid slugs. Another migration might be necessary for the `VersionAlias` model when it's actively used at the moment.

Otherwise the PR is ready for final review.

**Comment by ericholscher** [comment]
Yea, the VersionAlias is just available from the admin. It is a stub that I
wanted to build out to let people create Aliases for versions, so that they
can give them other names in RTD, since the tag name isn't always the most
useful for users.

On Mon, Jun 22, 2015 at 5:39 AM, Gregor Müllegger <notifications@github.com>
wrote:

> I added VersionSlugField and the migration necessary to normalize
> currently invalid slugs. Another migration might be necessary for the
> VersionAlias model when it's actively used at the moment.
>
> Otherwise the PR is ready for final review.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rtfd/readthedocs.org/pull/1362#issuecomment-114051458>
> .
>



-- 
Eric Holscher
Maker of the internet residing in Portland, Oregon
http://ericholscher.com

File: readthedocs/rtd_tests/tests/test_version_slug.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        version = Version.objects.create(
+            verbose_name='1%0',
+            project=self.pip)
+        self.assertEqual(version.slug, '1-0')
+
+    def test_normalizing_slashes(self):
+        version = Version.objects.create(
+            verbose_name='releases/1.0',
+            project=self.pip)
+        self.assertEqual(version.slug, 'releases__1.0')
```
Comment (ericholscher) [inline readthedocs/rtd_tests/tests/test_version_slug.py]: Still not sure about making ugly slugs by default here. I think we could pretty safely creates ``releases-1.0`` here, which is much nicer to read. I'm guessing we don't have a huge number of slugs that will have collisions between "/" and "-" in their names in the same repo.

We could do a query on prod, but my guess is there are 0 matching branch names that collide.

Comment (gregmuellegger) [inline readthedocs/rtd_tests/tests/test_version_slug.py]: Ok, I adjusted this in e51e5cc.

**Comment by gregmuellegger** [comment]
@ericholscher can the `VersionAlias` die then? We could at least move it into a branch. I'm happy to take the work.

**Comment by ericholscher** [comment]
Yea, we only have 4 defined in the DB, which were in place to help Django &
Fabric migrate. So we can likely kill them. Though it might be better to
just finish the feature and release it instead of killing it :)

On Mon, Jun 22, 2015 at 3:03 PM, Gregor Müllegger <notifications@github.com>
wrote:

> @ericholscher <https://github.com/ericholscher> can the VersionAlias die
> then? We could at least move it into a branch. I'm happy to take the work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rtfd/readthedocs.org/pull/1362#issuecomment-114222528>
> .
>



-- 
Eric Holscher
Maker of the internet residing in Portland, Oregon
http://ericholscher.com

**Comment by gregmuellegger** [comment]
True :D I will prepare something.

**Comment by ericholscher** [comment]
Think it makes sense to push that work onto another PR/Issue though, and get this stuff merged in.

**Comment by gregmuellegger** [comment]
I found another place that needs adjustment in this PR before it can be merged:
https://github.com/rtfd/readthedocs.org/blob/b4c6c7bcd750bd0751c7a151b485b22df932a435/readthedocs/builds/utils.py#L58

**Comment by gregmuellegger** [comment]
Argh, many other places use the slug as well:
https://github.com/rtfd/readthedocs.org/blob/b4c6c7bcd750bd0751c7a151b485b22df932a435/readthedocs/doc_builder/backends/sphinx.py#L103
https://github.com/rtfd/readthedocs.org/blob/b4c6c7bcd750bd0751c7a151b485b22df932a435/readthedocs/doc_builder/backends/sphinx.py#L122

**Comment by gregmuellegger** [comment]
Ok, this was hard work. Here it is ready for review.

**Comment by ericholscher** [comment]
Yea, that is actually a funny story. We historically allowed project slugs to have underscores, but that is invalid as a subdomain, so we did a bunch of hackery to fix them. We never updated the database to fix the existing slugs, because it might break a subset of pages. We should really do that :)

File: readthedocs/doc_builder/backends/sphinx.py [hunk 3]
```diff
@@ -83,7 +83,7 @@ def append_conf(self, **kwargs):
             display_bitbucket = True
 
         rtd_ctx = Context({
-            'current_version': self.version.slug,
+            'current_version': self.version.verbose_name,
```
Comment (ericholscher) [inline readthedocs/doc_builder/backends/sphinx.py]: Hmm, this should probably use the slug at some point in the future, but this is likely correct for backwards compat.

**Pull request merged**